### PR TITLE
 Support for transformers v5 post_process functions

### PR DESCRIPTION
**Pull request by shaddu**
# Description

`post_process_segmentation` is getting depreciated in transformers version 5. Code changes added are to support `post_process_semantic_segmentation` 

## Type of change

Please delete options that are not relevant.

-   [ ] New feature (non-breaking change which adds functionality)

## How has this change been tested, please provide a testcase or example of how you tested the change?

[Detection test Colab](https://colab.research.google.com/drive/12SdpvYtndq6bPPRquxYoMsyC9lkDEvvR?usp=sharing)

## Any specific deployment considerations

No

### CODE REVIEW

**Comment by SkalskiP** [comment]
Hi @shaddu 👋🏻 Thank you for submitting the PR. I have a few questions regarding the scope of the changes that have been and will need to be made in `from_transformers` to fully migrate us. In https://github.com/roboflow/supervision/issues/1113 I see that we need to migrate `post_process`, `post_process_panoptic`, `post_process_segmentation` and `post_process_instance`. Do I understand correctly that this PR only migrates `post_process_segmentation` for now, and the remaining outputs still need to be migrated?

**Comment by shaddu** [comment]
Hi @SkalskiP , Thank you for reviewing the PR and providing your comments. The scope was to migrate `post_process`, `post_process_panoptic`, `post_process_segmentation` and `post_process_instance`  but I have recently started working on computer vision, so I wanted to make sure before making all the changes that I am in the right direction. If this code is fine then I can migrate the rest of the method in this PR only or another as you prefer.

**Comment by SkalskiP** [comment]
@shaddu I like what you have done so far! Both code quality and functionality look good to me. From my perspective, the most important thing is that `from_transformers` supports both v4 and v5.

I think you can continue the migration in this PR.

**Comment by shaddu** [comment]
Hi @SkalskiP , 

I have finished adding support for the remaining function. Could you please review it and offer feedback?

Thank you

File: supervision/detection/core.py [hunk 1]
```diff
@@ -483,11 +484,40 @@ class names. If provided, the resulting Detections object will contain
             Class names values can be accessed using `detections["class_name"]`.
```
Comment (SkalskiP) [inline supervision/detection/core.py:484]: We need to update the `from_transformers` docstring. 

>  Creates a Detections instance from object detection or segmentation [Transformer](https://github.com/huggingface/transformers) inference result.

I'd mention we support object detection as well as panoptic, semantic and instance segmentation results.

> transformers_results (dict): The output of Transformers model inference. A dictionary containing the `scores`, `labels`, `boxes` and `masks` keys.

Now it can also be a `Tensor`.

**Comment by LinasKo** [comment]
Hi @shaddu

Thank you for your contribution! I'll check this today, hopefully we can merge soon.

File: supervision/detection/core.py [hunk 2]
```diff
@@ -483,11 +484,40 @@ class names. If provided, the resulting Detections object will contain
...
         """  # noqa: E501 // docs
 
-        class_ids = transformers_results["labels"].cpu().detach().numpy().astype(int)
         data = {}
```
Comment (SkalskiP) [inline supervision/detection/core.py:487]: In `from_transformers`, there are four places where we do:

```python
if id2label is not None:
    class_names = np.array([id2label[class_id] for class_id in class_ids])
    data[CLASS_NAME_DATA_FIELD] = class_names
```

I'd wrap it in a local helper function. It can be defined inside `from_transformers`.

```python
def get_data(class_ids: np.ndarray, id2label: Optional[Dict[int, str]]) -> dict:
    data = {}
    if id2label is not None:
        class_names = np.array([id2label[class_id] for class_id in class_ids])
        data[CLASS_NAME_DATA_FIELD] = class_names
    return data
```

File: supervision/detection/core.py [hunk 3]
```diff
@@ -504,6 +534,60 @@ class names. If provided, the resulting Detections object will contain
...
+                    )
+                    data[CLASS_NAME_DATA_FIELD] = class_names
+
+                return cls(
+                    xyxy=mask_to_xyxy(masks),
+                    mask=masks,
+                    class_id=class_ids,
+                    data=data,
+                )
         else:
```
Comment (SkalskiP) [inline supervision/detection/core.py:591]: I think we can drop this `else` statement.

File: supervision/detection/core.py [hunk 4]
```diff
@@ -483,11 +484,40 @@ class names. If provided, the resulting Detections object will contain
...
 
-        class_ids = transformers_results["labels"].cpu().detach().numpy().astype(int)
         data = {}
-        if id2label is not None:
-            class_names = np.array([id2label[class_id] for class_id in class_ids])
-            data[CLASS_NAME_DATA_FIELD] = class_names
+
+        if transformers_results.__class__.__name__ == "Tensor":
+            segmentation_array = transformers_results.cpu().detach().numpy()
+
```
Comment (SkalskiP) [inline supervision/detection/core.py:491]: Let's make it more compact and drop those extra new lines.

File: supervision/detection/utils.py [hunk 5]
```diff
@@ -1000,3 +1002,20 @@ def cross_product(anchors: np.ndarray, vector: Vector) -> np.ndarray:
...
     vector_start = np.array([vector.start.x, vector.start.y])
     return np.cross(vector_at_zero, anchors - vector_start)
+
+
+def png_to_mask(png_string):
+    """
+    Convert a PNG byte string to a binary mask array.
+
+    Args:
+    - png_string (bytes): A byte string representing the PNG image.
```
Comment (SkalskiP) [inline supervision/detection/utils.py:1012]: This docstring format is incompatible with rest of docstings in the project.

File: supervision/detection/utils.py [hunk 6]
```diff
@@ -1000,3 +1002,20 @@ def cross_product(anchors: np.ndarray, vector: Vector) -> np.ndarray:
...
+
+    Returns:
+    - np.ndarray: A binary mask array with shape (H, W), where H and W
+        are the height and width of the image.
+    """
+    image = Image.open(io.BytesIO(png_string))
+    mask = np.array(image, dtype=np.uint8)
+
```
Comment (SkalskiP) [inline supervision/detection/utils.py:1020]: Let's drop this extra new line.

File: supervision/detection/core.py [hunk 7]
```diff
@@ -504,6 +534,60 @@ class names. If provided, the resulting Detections object will contain
                 class_id=class_ids,
                 data=data,
             )
+        elif "segments_info" in transformers_results:
```
Comment (SkalskiP) [inline supervision/detection/core.py:537]: Personally, I would move `segments_info = transformers_results["segments_info"]` extraction to ` if "segmentation" in transformers_results:` and `elif "png_string" in transformers_results:` blocks. We will get one indention less.

**Review by SkalskiP — CHANGES_REQUESTED** [review]

**Comment by SkalskiP** [comment]
Hi @shaddu 👋🏻 Awesome work here. I left some comments regarding the current version of the code, but here are some high-level overviews ones:
- Could you update your Colab? The current version tests only a limited subset of supported capabilities. It would be awesome to test all functions mentioned in https://github.com/roboflow/supervision/issues/1113 in a single Colab.
- In my opinion, having all that logic in a single function is hard to follow. I would create a new file, `supervision/detection/transformers.py`, and inside define `process_transformers_v5_result`, `process_transformers_v4_detection_result`, `process_transformers_v4_segmentation_result`, etc, and just call those inside `from_transformers` in `supervision/detection/core.py`. This will:
    - Make `supervision/detection/core.py` shorter. It is already too long.
    - Make transformers' logic organized and easier to maintain. 

In general this looks really good already. Thanks a lot for all the help! 🙏🏻

**Comment by LinasKo** [comment]
@shaddu, @SkalskiP, I'll help out a little bit - I'm halfway towards making a more well-rounded Colab

**Comment by shaddu** [comment]
Hey @SkalskiP, 

Thank you for the detailed feedback. The suggestion of a separate file is a good idea. I'll work on this approach.

Hey @LinasKo,

Thank you for sharing the new collab. I learned something new today on how to use your repos in collab notebook 👍

File: supervision/detection/core.py (hunk 1 above)
Comment (shaddu) [reply supervision/detection/core.py:490]: Hi @SkalskiP , One quick query, ideally it should now be `transformers_results (Union[dict, torch.Tensor])` but to support this we need to import the torch package, should we add the import or leave it in docstrings for now?

**Comment by shaddu** [comment]
Hi @SkalskiP / @LinasKo  , 

I have made changes as per the feedback, can you please review the PR again?

@LinasKo  `post_process_panoptic_segmentation` adds class names in `detection.data`. 

I will be adding more models and test cases in this [colab](https://colab.research.google.com/drive/1YYvWeAxeiwXaHp-2InHyk1UZUIfyk80-#scrollTo=HvKs-upVUfZV) for review.

**Comment by SkalskiP** [comment]
Hi @shaddu 👋🏻 ! Awesome work! I added small changes - changed names of functions and improved docstrings. We are ready to merge!

**Comment by shaddu** [comment]
@SkalskiP Thank you for merging the PR. It was a great learning experience. I'll look for more issues to work on in Supervision and if you come across any related issues, I'd be happy to help with them.

**Comment by SkalskiP** [comment]
@shaddu, working with you was an awesome experience! 🔥
